### ISSUE TITLE
Activating gradient checkpointing

### ISSUE THREAD

**Issue by ShivanshuPurohit** [huggingface/transformers#11448]
Working on a script which uses `GPTNeoForCausalLM` as the model object. As I understand it, gradient checkpointing requires checkpointing every layer. How do I change `GPTNeoForCausalLM` to incorporate gradient checkpointing, given that it doesn't show the layers explicitly but rather uses
```
self.transformer = GPTNeoModel(config)
self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)
```
PS. working on [visua-grounding](https://github.com/EleutherAI/visual-grounding/tree/main) with which @stas00 has already been a huge help, and knows the issues.

**Comment by stas00** [comment]
Unless I'm missing something I haven't heard of "gradient checkpointing". There is are "activation checkpointing" and "gradient accumulation". The latter is automatic, the former indeed requires changes to the code. Which one are you after?

**Comment by ShivanshuPurohit** [comment]
Ah. Thanks for the clarification.
